File: app/src/main/java/com/aistra/hail/utils/AppMetadataDao.java
package com.aistra.hail.utils;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

@Dao
public interface AppMetadataDao {
    @Query("SELECT * FROM app_metadata")
    List<AppMetadataEntity> loadAll();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void upsertAll(List<AppMetadataEntity> entries);

    @Query("DELETE FROM app_metadata")
    void deleteAll();

    @Query("UPDATE app_metadata SET installed = 0")
    void markAllUninstalled();

    @androidx.room.Transaction
    default void replaceAll(List<AppMetadataEntity> entries) {
        deleteAll();
        upsertAll(entries);
    }
}

File: app/src/main/kotlin/com/aistra/hail/utils/AppMetadataDao.kt
package com.aistra.hail.utils

import androidx.room3.Dao
import androidx.room3.Query
import androidx.room3.Transaction
import androidx.room3.Upsert

@Dao
interface AppMetadataDao {
    @Query("SELECT * FROM app_metadata")
    fun loadAll(): List<AppMetadataEntity>

    @Upsert
    fun upsertAll(entries: List<AppMetadataEntity>)

    @Query("DELETE FROM app_metadata")
    fun deleteAll()

    @Query("UPDATE app_metadata SET installed = 0")
    fun markAllUninstalled()

    @Transaction
    fun replaceAll(entries: List<AppMetadataEntity>) {
        deleteAll()
        upsertAll(entries)
    }
}
